#806 Disabling BSF-Resources doesn´t work

cljk opened the pull request:
Addressing issue #806 to be able to load own version of Bootstrap - like described in https://showcase.bootsfaces.net/miscellaneous/Configuration.jsf ("Use your own version of jQuery, jQueryUI, FontAwesome, or Bootstrap instead of using the version BootsFaces brings").

Demo that fix is working: https://github.com/cljk/bsf-issue-806-example

Open issue: CSS of components still get added (perhaps to be fixed later). bsf.css is still added - which should be correct (and modified in the documentation).


Changes:
* move invoking removeBootstrapResources() to end of addCSS() (CSS files were removed and added again)
* fixed logic of "load Bootstrap from CDN" in addCSS() and renamed flag (compare to load jQuery logic)
* added position "last" to "bsf.js" to be able to add custom jQuery before (bsf.js depends on it)
* added named constants for paremeters

stephanrauh commented:
You mentioned you've renamed the flag `net.bootsfaces.get_bootstrap_from_cdn`.  However, as far as I can see, the name of the flag is the same, but you've inverted its meaning. Did I miss something?

TheCoder4eu commented:
@cljk , @stephanrauh , the statement 
`loadBootstrap = shouldLibraryBeLoaded(P_GET_BOOTSTRAP_FROM_CDN, true);`

assumes by default it is true we want to load Bootstrap from CDN instead than from the library.
For this reason I liked more the name of the variable loadBootstrap to be loadBootstrapFromCDN.
Hence the statement 
`if (loadBootstrapFromCDN) { removeBootstrapResources(root, context); }`
made perfectly sense, so I agree the semantic of the parameter is inverted in the PR.

stephanrauh commented:
I don't mind the inversion as such - I'm just concerned about backward compatibility. Does this change break existing programs? If so, we should call the next version of BootsFaces 1.2.0, not 1.1.2. Plus, we should shout the breaking change out load and clear.

cljk commented:
Sorry for the late reply ... 

* I didn´t rename the parameter `net.bootsfaces.get_bootstrap_from_cdn` - but the flag from `loadBootstrapFromCDN` to `loadBootstrap` - because that is what the logic result of `shouldLibraryBeLoaded` actually is. The old name was misleading and source for the adressed bug. You can compare that to the other parameters like `loadJQuery` or `loadJQueryUI` which are used the same (and are not changed). So the *meaning* of the flag is the opposite as before - and it always was.... but... 
* it was non-functional.... because the logic was removing the Bootstrap-CSS .... but this was done in the wrong order and they were added again. Because of that I moved the block to lines 308 ff. It was also non-functional because the resources were removed if `loadBootstrapFromCDN` was true - which was (see above) actually the wrong interpretation of the flag and the reason why I renamed it. So the resources are removed if `! loadBootstrap`.
* extracting the parameter-names to the constant-class C was just cosmetical

Since this patch just enables the feature `net.bootsfaces.get_bootstrap_from_cdn` described in your own doc https://showcase.bootsfaces.net/miscellaneous/Configuration.jsf I don´t think this should break any application. But thats a decision to you.

cljk commented:
Thank you! ;-)
